text/0000-futures.md

```diff
@@ -0,0 +1,811 @@
+is a final step in standardization.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `Future` trait represents an *asynchronous* computation that may eventually
+produce a final value, but don't have to block the current thread to do so.
+
+Futures are usually constructed via *asynchronous functions*:
+
+```rust
+async fn read_frame(socket: &TcpStream) -> Result<Frame, io::Error> { ... }
+```
+
+This `async` function, when invoked, produces a *future* that represents the
+completion of reading a frame from the given socket. The function signature
+is equivalent to:
+
+```rust
+fn read_frame<'sock>(socket: &'sock TcpStream)
+    -> impl Future<Item = Result<Frame, io::Error>> + 'sock;
+```
+
+Other async functions can *await* this asynchronous value; see the [companion
+RFC] for details.
+
+In addition to `async fn` definitions, futures can be built using adapters on
+the `Future` trait, much like with `Iterator`s. The standard library includes a
+number of basic adapters (described in the reference below), while some
+particularly interesting variants are iterating in the crates.io ecosystem
+first.
+
+Ultimately asynchronous computations are executed by *tasks*, which are
+lightweight threads. In particular, an *executor* is able to "spawn" a
+`()`-producing future as an independent task; these tasks are then cooperatively
+scheduled onto one or more operating system threads. The `Executor` trait
+defines this interface, and the `task` module provides a host of related
+definitions needed when manually implementing futures or executors.
+
+*Note: additional guide-level documentation is available in
+the [futures crate](https://docs.rs/futures/0.2.0-beta/futures/)*.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `core::task` module
+
+The fundamental mechanism of futures is *tasks*, which are lightweight
+threads of execution; many tasks can be cooperatively scheduled onto a single
+operating system thread.
+
+To perform this cooperative scheduling we use a technique sometimes referred to
+as a "trampoline". When a task would otherwise need to block waiting for some
+event, instead it schedules itself for later wakeup and *returns* to the
+executor running it, which can then run another task. Subsequent wakeups place
+the task back on the executors queue of ready tasks, much like a thread
+scheduler in an operating system.
+
+Attempting to complete a task (or future within it) is called *polling*, and
+always yields a `Poll` value back:
+
+```rust
+/// Indicates whether a value is available, or if the current task has been
+/// scheduled for later wake-up instead.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Poll<T> {
+    /// Represents that a value is immediately ready.
+    Ready(T),
+
+    /// Represents that a value is not ready yet.
+    ///
+    /// When a function returns `Pending`, the function *must* also
```

Comment:
An alternative design that should be at least mentioned in the alternatives section is to make `Poll::Pending` wrap some "I have scheduled my task!" token, which a future can obtain either from `Poll::Pending` being returned from one of its children futures or by explicitly constructing such a token (using some method/API whose naming would make it painfully clear it's not okay to just call it without actually scheduling the task to be woken up by some external mechanism), yay learnability, yay typesystem. This was proposed (and, I assume, rejected) before, so it should make sense to link to that discussion from this RFC.

Reply:
I agree, it is really nice when strong wording in comments can be replaced with a little type hack which ensures that the API cannot be used incorrectly.

Reply:
Well, incorrect usage is still possible if you have multiple children futures and forget to poll some of them,  but seems much less likely.